manuscript/13-Modules.md

```diff
@@ -456,13 +456,186 @@ This example loads `module.js` as a module instead of a script by passing a seco
+
+```js
+import { sum } from "./example";
+```
+
+When Node.js sees `"./example"`, a path without an extension, it searches the following locations in order, moving on to the next location if the previous one isn't found:
+
+1. `./example.mjs`
+1. `./example.js`
+1. `main` entry in `./example/package.json`
+1. `./example/index.mjs`
+1. `./example/index.js`
+
+This lookup sequence allows you to place JavaScript modules and CommonJS modules in the same directory, ensuring (but not requiring) backwards compatibility for other consumers. Those using an older version of Node.js would always load the `.js` file whereas a newer version would load the`.mjs` file. The lookup sequence is particular important in the case of packages installed in the `node_modules` directory, such as:
+
+```js
+import { sum } from "example";
+```
+
+In this case, `"example"` is a package name rather than a filename. As such, the lookup sequence is slightly different:
+
+1. `./node_modules/example.mjs`
+1. `./node_modules/example.js`
+1. `main` in `./node_modules/example/package.json`
+1. `./node_modules/example/index.mjs`
+1. `./node_modules/example/index.js`
+1. Repeat this process in the parent directory
+
+As with single files, packages can contain both `.mjs` and `.js` files to ensure compatibility with older versions of Node.js.
+
+A> #### Module Resolution Differences Between the Browser and Node.js
+A>
+A> The ECMAScript 6 standard does not specify the format of module filenames in `import` statements, leaving those details up to the implementations. As such, there's a subtle but important difference in the way browsers and Node.js resolve modules. Consider the following:
+A>
+A> ```js
+A> import { sum } from "example/sum.js";
+A> ```
+A>
+A> While this code may look simple to understand, browsers and Node.js interpret `"example/sum.js"` differently. The differences are outlined in the table below.
+A>
+A> | **Environment** | **File to Load** | **Loaded As** | **Mode** |
+A> -----------------------------------------------------------------
+A> | Browser         | `./example/sum.js` | JavaScript module | strict |
+A> | Node.js         | `./node_modules/example/sum.js` | CommonJS module | nonstrict |
+A>
+A> In a browser, `"example"` is resolved as a subdirectory whereas Node.js resolves `"example"` as a package name. Further, because the `.js` file extension is specified (which is required by the browser), the browser loads the file as a JavaScript module while Node.js loads the file as CommonJS. These differences pose some interoperability concerns for those who want to write JavaScript modules that work in both the browser and Node.js. The upcoming module loader specification is intended to help solve some of these module resolution interoperability concerns.
```

Comment:
Note: browsers are not determining the type of file by extension, they only use the MIME in the content-type header, this MIME is shared for Script and Module goal of the language; it just assumes Module. This assumption has a grammatical ambiguity which means things meant for Script run in the wrong mode ([and parse/run, but produce incorrect values](https://gist.github.com/bmeck/e54d607a4b0874e8dfca7af2a132b9c2))...

Its pretty easy to see this in action by loading something like `<script src=my.php>` with any JS MIME in https://html.spec.whatwg.org/multipage/scripting.html#scriptingLanguages